src/dotnet/ShellShim/ShimCreationTransaction.cs

```diff
@@ -0,0 +1,42 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Transactions;
+
+namespace Microsoft.DotNet.ShellShim
+{
+    public class ShimCreationTransaction : IEnlistmentNotification
```

Comment:
Could `ShimCreationTransaction` and `ObtainTransaction` perhaps be generalized?  They seem to be near duplicates of one another, except that the latter has a  rollback action, which I think is missing from shim creation (rollback should delete the shim files; for example, what happens if we create the `.exe` but fail to create the `.config` file, or vice versa?).

On a side note: I like this use of the `TransactionScope` and `IEnlistmentNotification`.

Reply:
Generalize `ShimCreationTransaction`, `ObtainTransaction` -- they are very likely could,  it is almost java's anonymous class. but i cannot find a good place to put the shared code. Any suggestion?

ShimCreationTransaction is treaty. In the beginning, i put createshim as a separate method, and roll back is delete. And then i realize rollback is deleting the existing shim (the conflicted one). Then I realize that is wrong, it need 2 stage. What about put in .stage and Directory.move.? -- the whole shim creation is a Directory.Move. So I just move them to commit and add verification in prepare.

At the same time, it fits this guideline very well. https://stackoverflow.com/questions/21094406/where-should-i-perform-action-when-implementing-ienlistmentnotification